search/scorer/scorer_knn.go

```diff
@@ -72,19 +70,50 @@ func (sqs *KNNQueryScorer) Score(ctx *search.SearchContext,
 	if sqs.includeScore || sqs.options.Explain {
 		var scoreExplanation *search.Explanation
 		score := knnMatch.Score
-		// in case of euclidean distance being the distance metric,
-		// an exact vector (perfect match), would return distance = 0
-		if score == 0 {
-			score = maxKNNScore
-		} else {
-			// euclidean distances need to be inverted to work with
-			// tf-idf scoring
-			score = 1.0 / score
+		if sqs.similarityMetric == index.EuclideanDistance {
+			// in case of euclidean distance being the distance metric,
+			// an exact vector (perfect match), would return distance = 0
+			if score == 0 {
+				score = maxKNNScore
+			} else {
+				// euclidean distances need to be inverted to work with
+				// tf-idf scoring
+				score = 1.0 / score
+			}
+		}
+
+		if sqs.options.Explain {
+			childrenExplanations := make([]*search.Explanation, 1)
+			childrenExplanations[0] = &search.Explanation{
+				Value: score,
+				Message: fmt.Sprintf("vector(field(%s:%s) with similarity_metric(%s)=%e",
+					sqs.queryField, knnMatch.ID, sqs.similarityMetric, score),
+			}
+			scoreExplanation = &search.Explanation{
+				Value: score,
+				Message: fmt.Sprintf("fieldWeight(%s in doc %s), score of:",
+					sqs.queryField, knnMatch.ID),
+				Children: childrenExplanations,
+			}
 		}
 
 		// if the query weight isn't 1, multiply
 		if sqs.queryWeight != 1.0 && score != maxKNNScore {
 			score = score * sqs.queryWeight
+			if sqs.options.Explain {
+				childExplanations := make([]*search.Explanation, 2)
```

Comment:
It seems we're double allocating unnecessarily above and below, this should be within an if and else.

Reply:
@metonymic-smokey address this separate please.